(window.webpackJsonp=window.webpackJsonp||[]).push([[497],{2363:function(r,t,n){"use strict";n.r(t);var l=n(46),e=Object(l.a)({},(function(){var r=this,t=r.$createElement,n=r._self._c||t;return n("ContentSlotsDistributor",{attrs:{"slot-key":r.$parent.slotKey}},[n("h2",{attrs:{id:"try-catch-finally和return的执行顺序"}},[n("a",{staticClass:"header-anchor",attrs:{href:"#try-catch-finally和return的执行顺序"}},[r._v("#")]),r._v(" try-catch-finally和return的执行顺序")]),r._v(" "),n("ul",[n("li",[n("p",[r._v("try{}catch(){}finally{} return")]),r._v(" "),n("p",[r._v("程序按顺序执行")])]),r._v(" "),n("li",[n("p",[r._v("try{return;}catch(){}finally{} return;")]),r._v(" "),n("p",[r._v("程序执行try块中return之前(包括return语句中的表达式运算)代码")]),r._v(" "),n("p",[r._v("在执行finally块，最后执行try中return")]),r._v(" "),n("p",[r._v("finally块之后的语句return，因为程序在try中已经return所以不在执行")])]),r._v(" "),n("li",[n("p",[r._v("try{}catch(){return} finally{} return")]),r._v(" "),n("p",[r._v("程序先执行try，如果遇到异常执行catch块")]),r._v(" "),n("p",[r._v("有异常:则执行catch中return之前(包括return语句中的表达式运算)代码，在执行finally语句中全部代码，最后执行catch块中return,finally之后也就是4处的代码不在执行")]),r._v(" "),n("p",[r._v("无异常:执行完try在finally在return")])]),r._v(" "),n("li",[n("p",[r._v("try{return} catch(){} finally{return}")]),r._v(" "),n("p",[r._v("程序执行try块中return之前(包括return语句中表达式运算)代码")]),r._v(" "),n("p",[r._v("在执行finally块，因为finally块中有return所以提前退出")])]),r._v(" "),n("li",[n("p",[r._v("try{}catch(){return;}finally{return}")]),r._v(" "),n("p",[r._v("程序执行catch块中return之前(包括return语句中的表达式运算)代码;")]),r._v(" "),n("p",[r._v("在执行finally块，因为finally块中有return所以提前退出")])]),r._v(" "),n("li",[n("p",[r._v("try{return;}catch() {return;}finally{return;}")]),r._v(" "),n("p",[r._v("程序执行try块中return之前(包括return语句中不表达式运算);")]),r._v(" "),n("p",[r._v("有异常:执行catch块中return之前(包括return语句中的表达式运算)代码；则在执行finally块，因为finally快中有return所以提前退出")]),r._v(" "),n("p",[r._v("无异常:则在执行finally块，因为finally块中有return所以提前退出")])])]),r._v(" "),n("p",[n("span",{staticStyle:{color:"red"}},[r._v("最终结论："),n("strong",[r._v("任何执行try或者catch的return语句之前，都会先执行finally语句，如果finally存在的话。如果finally中有return语句，那么程序就return，所以finally中的return是一定会被return的")])])]),r._v(" "),n("h2",{attrs:{id:"资料"}},[n("a",{staticClass:"header-anchor",attrs:{href:"#资料"}},[r._v("#")]),r._v(" 资料")]),r._v(" "),n("p",[n("a",{attrs:{href:"https://www.cnblogs.com/yanbigfeg/p/9295541.html",target:"_blank",rel:"noopener noreferrer"}},[r._v("try、catch、finally详解，你不知道的异常处理 "),n("OutboundLink")],1)]),r._v(" "),n("p",[n("a",{attrs:{href:"https://www.cnblogs.com/qianmengzbx/p/14531502.html",target:"_blank",rel:"noopener noreferrer"}},[r._v("try-catch-finally和return的执行顺序"),n("OutboundLink")],1)])])}),[],!1,null,null,null);t.default=e.exports}}]);